### DISCUSSION TITLE
Build using environment variables in staging

### DISCUSSION THREAD

**Discussion by AdagioB** [quasarframework/quasar#17191] [General - Components / Directives / etc] [answered]
I'm trying to learn, what is the correct way to use environment variables, that changes based on the mode of the app
Unfortunately I'm having a hard time finding a lot of useful information, that can help me with my case. This is a bit weird, as I believe this is a very common scenario

The way I see it, I should have three different modes: development, staging and production

At first I thought I would need to edit quasar.config.js to something like this

`build: {
      env: {
        API: ctx.dev
          ? 'https://dev.api.com'
          : ctx.staging
            ? 'https://test.api.com'
            : 'https://prod.api.com'
      }`

But as far as I can tell, there's only ctx.dev and ctx.production

Then I tried to add the files .env.production, .env.staging and .env where I added VITE_API= with their respective addresses, hoping for some magic here. Unfortunately I have no idea how to build it for staging

`"scripts": {
	"lint": "eslint --ext .js,.ts,.vue ./",
	"format": "prettier --write \"**/*.{js,ts,vue,,html,md,json}\" --ignore-path .gitignore",
	"test": "quasar build",
	"dev": "quasar dev",
	"build": "quasar build"
},`

Here I run into the same problem. I can only find commands to build in production or development
This leaves me thinking that apparently nobody sets up some test servers anymore to test their stuff... real men test in production... or am I missing something here?

**Comment by AdagioB** [comment, marked as answer]
I finally managed to find a workaround to get it working

I created this file: \variables\parser.js

`const dotenv = require('dotenv');

const files = {
	...dotenv.config({ path: 'variables/.env' }).parsed,
	...dotenv.config({ path: `variables/.env.${process.env.ENVIRONMENT}` }).parsed,
	...dotenv.config({ path: `variables/.env.${process.env.ENVIRONMENT}.local` }).parsed,
};

module.exports = () =>
{
	Object.keys(files, (key) =>
	{
		if (typeof files[key] !== 'string')
		{
			files[key] = JSON.stringify(files[key]);
		}
	});
	return files;
};
`

In the same folder I created different .env files (e.g. .env.staging) where I added the variables they needed

Inside quasar.config.js I added this line to the top:
`const PARSER = require('./variables/parser')();`

And further down I did this:
`build: {
env: PARSER
}`

In package.json I added this line unders scripts:

`"staging": "cross-env ENVIRONMENT=staging quasar build -m spa",`

I found the information in this discussion: https://github.com/quasarframework/quasar/discussions/15065

**Reply by rstoenescu** [reply]
You could use the q/app-vite v2.0.0-beta or q/app-webpack v4.0.0-beta. Both have native support for [dotenv files](https://quasar.dev/quasar-cli-vite/upgrade-guide#the-env-dotfiles-support)

**Reply by true-vue** [reply]
@rstoenescu thanks!

Using [vite build-in way](https://vitejs.dev/guide/env-and-mode.html) for handling .env files is great feature. Separate one that (still) seems to be missing is ability to pass vite mode into quasar cli command.

So something like: `quasar build --vite-mode staging`

Is it considered to be done soon?

**Reply by ptc-tonyo** [reply]
Sadly `cross-env ENVIRONMENT=staging quasar build -m spa` does not appear to work.

**Comment by ptc-tonyo** [comment]
I found that this works in Quasar CLI 2.5.0 . Inside `quasar.conf.js` in the `build` section you can set Vite's mode like so:
This will load env variables from `.env.staging` as per Vite documentation . 

```js
extendViteConf(viteConf) {
  if (process.env.ENVIRONMENT === 'staging') {
     viteConf.mode = 'staging'
  }
},
```

_Warning_: This is a bit of a **hack** and thus risky as we don't know what effect it will have on Quasar's build rules either now or in a future version, since it's not officially supported by Quasar. Hence the request that setting the Vite mode from the CLI be an official option.
